psl-core/src/main/java/org/linqs/psl/application/util/GroundRules.java

```diff
@@ -193,6 +194,123 @@ public static double getExpectedWeightedCompatibility(WeightedGroundRule groundR
+	 * with probability equal to its current truth value.
+	 *
+	 * WARNING: the result of this function is incorrect if the RandomVariableAtoms
+	 * are subject to any {@link UnweightedGroundRule UnweightedGroundRule}.
+	 *
+	 * WARNING: This method does not account for WeightedGroundLogicalRules that
+	 * were not grounded because they are trivially satisfied.
+	 */
+	public static double getExpectedWeightedLogicalCompatibility(WeightedGroundLogicalRule groundRule) {
+		/* Collects GroundAtom values. Atoms in GroundRule are
+		 * already negated */
+		List<Double> posRvValues = new ArrayList<Double>();
+		List<Double> posObsValues = new ArrayList<Double>();
+		List<Double> negRvValues = new ArrayList<Double>();
+		List<Double> negObsValues = new ArrayList<Double>();
+		for (GroundAtom atom : groundRule.getPositiveAtoms())
+			if (atom instanceof RandomVariableAtom)
+				negRvValues.add(atom.getValue());
+			else
+				negObsValues.add(atom.getValue());
+		for (GroundAtom atom : groundRule.getNegativeAtoms())
+			if (atom instanceof RandomVariableAtom)
+				posRvValues.add(atom.getValue());
+			else
+				posObsValues.add(atom.getValue());
+
+		// Sum of observed atoms
+		double k = 0;
+		for (double posObsValue : posObsValues)
+			k += posObsValue;
+		for (double negObsValue : negObsValues)
+			k += (1 - negObsValue);
+		k = Math.min(1, k);
+
+		// P(sum of RV >= 1)
+		double notp = 1.0;
+		for (double posRvValue: posRvValues)
+			notp *= 1 - roundingProbability(posRvValue);
+		for (double negRvValue: negRvValues)
+			notp *= roundingProbability(negRvValue);
+		double p = 1.0 - notp;
+
+		// 2 cases: If p, then compatibility = 1, else = k.
+		double compat = p + notp * k;
+		return groundRule.getWeight() * compat;
+	}
+
+	/**
+	 * Computes the weighted probability WeightedGroundLogicalRule is satisfied.
+	 * from independently rounding each {@link RandomVariableAtom} to 1.0 or 0.0
+	 * with probability equal to its current truth value.
+	 *
+	 * WARNING: this function will throw an exception if any GroundAtoms are soft-valued.
+	 *
+	 * WARNING: the result of this function is incorrect if the RandomVariableAtoms
+	 * are subject to any {@link UnweightedGroundRule UnweightedGroundRule}.
+	 *
+	 * WARNING: This method does not account for WeightedGroundLogicalRules that
+	 * were not grounded because they are trivially satisfied.
+	 */
+	public static double getExpectedWeightedLogicalSatisfaction(WeightedGroundLogicalRule groundRule) throws Exception {
+
+		/* Collects GroundAtoms. Atoms in GroundRule are
+		 * already negated */
+		List<GroundAtom> posAtoms = new ArrayList<GroundAtom>();
+		List<GroundAtom> negAtoms = new ArrayList<GroundAtom>();
+		for (GroundAtom atom : groundRule.getPositiveAtoms())
+			if (!(atom instanceof RandomVariableAtom) &&
+				(atom.getValue() > 0) &&
+				(atom.getValue() < 1))
+				throw new Exception("Soft observed values are not supported");
```

Comment:
Throw a non-checked exception like an IllegalArgumentException.
Then you don't need to declare it as thrown.
Always avoid throwing checked exceptions when possible and appropriate.